src/lib.rs

```diff
@@ -1,7 +1,134 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+#![feature(iterator_fold_self)]
+
+pub mod cpu {
+    use std::io::{BufRead, BufReader, Read};
+
+    #[derive(Debug, PartialEq)]
+    pub struct Stat {
+        pub user: u64,
+        pub nice: u64,
+        pub system: u64,
+        pub idle: u64,
+        pub iowait: u64,
+        pub irq: u64,
+        pub softirq: u64,
+        pub steal: u64,
+        pub guest: u64,
+        pub guest_nice: u64,
+        pub total: u64,
+        pub cpu_count: i64,
+        pub stat_count: i64,
     }
+
+    impl Default for Stat {
+        fn default() -> Self {
+            Self {
+                user: 0,
+                nice: 0,
+                system: 0,
+                idle: 0,
+                iowait: 0,
+                irq: 0,
+                softirq: 0,
+                steal: 0,
+                guest: 0,
+                guest_nice: 0,
+                total: 0,
+                cpu_count: 0,
+                stat_count: 0,
+            }
+        }
+    }
+
+    pub fn get() -> std::io::Result<Stat> {
+        let file = std::fs::File::open("/proc/stat")?;
+        collect_cpu_stats(file)
+    }
+
+    pub fn collect_cpu_stats<R: Read>(file: R) -> std::io::Result<Stat> {
+        let mut reader = BufReader::new(file);
+        let mut line = String::new();
+        let mut total = 0;
+        reader.read_line(&mut line)?;
+
+        let mut iterator = line.split_ascii_whitespace();
+        // the first token is "cpu", so skip it
+        iterator.next();
+        let vals: Vec<_> = iterator
+            .map(|val| {
+                let val = val.parse::<u64>();
+                if val.is_ok() {
+                    val.unwrap()
+                } else {
+                    unimplemented!()
+                }
+            })
+            .collect();
+        // if vals.len() != 10 {
+        //     unimplemented!()
+        // }
+        let mut stat = Stat::default();
+        stat.user = vals[0];
+        stat.nice = vals[1];
+        stat.system = vals[2];
+        stat.idle = vals[3];
+        stat.iowait = vals[4];
+        stat.irq = vals[5];
+        stat.softirq = vals[6];
+        stat.steal = vals[7];
+        stat.guest = vals[8];
+        stat.guest_nice = vals[9];
```

Comment:
I'd like to come up with a good idea for improving the code around here, because it's very insecure about accessing the slices directly.